homeassistant/components/statistics/sensor.py

```diff
@@ -209,92 +227,105 @@ def _purge_old(self):
         if self._max_age is not None:
             self._purge_old()
 
         self.count = len(self.states)
 
         if not self.is_binary:
-            try:  # require only one data point
-                self.mean = round(statistics.mean(self.states), self._precision)
-                self.median = round(statistics.median(self.states), self._precision)
-            except statistics.StatisticsError as err:
-                _LOGGER.debug("%s: %s", self.entity_id, err)
-                self.mean = self.median = STATE_UNKNOWN
-
-            try:  # require at least two data points
-                self.stdev = round(statistics.stdev(self.states), self._precision)
-                self.variance = round(statistics.variance(self.states), self._precision)
-            except statistics.StatisticsError as err:
-                _LOGGER.debug("%s: %s", self.entity_id, err)
-                self.stdev = self.variance = STATE_UNKNOWN
-
-            if self.states:
-                self.total = round(sum(self.states), self._precision)
-                self.min = round(min(self.states), self._precision)
-                self.max = round(max(self.states), self._precision)
-
-                self.min_age = self.ages[0]
-                self.max_age = self.ages[-1]
-
-                self.change = self.states[-1] - self.states[0]
-                self.average_change = self.change
-                self.change_rate = 0
-
-                if len(self.states) > 1:
-                    self.average_change /= len(self.states) - 1
-
-                    time_diff = (self.max_age - self.min_age).total_seconds()
-                    if time_diff > 0:
-                        self.change_rate = self.average_change / time_diff
-
-                self.change = round(self.change, self._precision)
-                self.average_change = round(self.average_change, self._precision)
-                self.change_rate = round(self.change_rate, self._precision)
+            self._process_statistics()
 
-            else:
-                self.total = self.min = self.max = STATE_UNKNOWN
-                self.min_age = self.max_age = dt_util.utcnow()
-                self.change = self.average_change = STATE_UNKNOWN
-                self.change_rate = STATE_UNKNOWN
+        await self._set_expiration_update_for_max_age_sensors()
 
-        # If max_age is set, ensure to update again after the defined interval.
+    async def _set_expiration_update_for_max_age_sensors(self):
         next_to_purge_timestamp = self._next_to_purge_timestamp()
         if next_to_purge_timestamp:
-            _LOGGER.debug(
-                "%s: scheduling update at %s", self.entity_id, next_to_purge_timestamp
-            )
-            if self._update_listener:
+            if (
+                self._update_listener is not None
+                and self._last_scheduled_purge > dt_util.utcnow()
+            ):
+                # cancel scheduled update only if it is in the future
```

Comment:
cc @balloob 

This is the key to avoid the core warning.

The usual listener cancellation involves a listener set for some time in the _far_ future, so its cancellation and new schedule work always ok.

But, _sometimes_, the listener is set _in the past_ (:= it's already activated, or it's in the process to be).

In those cases, again _sometimes_, the async_update checks `self._update_listener` and sees it active, as `_scheduled_update` has not run yet.

In `helpers/event.py:async_track_point_in_utc_time`, this is happening:

```python
# This happened:
setattr(point_in_time_listener, "run", True)
async_unsub()  # so the listener is already removed

# But the action has not been called yet, it's in queue:
hass.async_run_job(action, now)
```

So when we call `self._update_listener()`, the core warning appears.

I was trying to solve the root causes, assuming that a lower number of updates could do the trick, but in the end, it is only to handle that particular moment in which the listener is caught in the middle.

I cannot assume that my particular use case can be extrapolated for anyone suffering #31997, but I think this plot is quite significant:

<img width="450" alt="Unknown" src="https://user-images.githubusercontent.com/9748301/75971304-e0263e80-5ed1-11ea-9842-ce8725d61688.png">
(sorry for the messy styling)


That shows the scheduled update cancel operations, where _y:=ts_sched - now (s)_. Almost all canceled listeners are scheduled in ~500 seconds in the future, but there are some points where ∆T is ~0 (negative value around ~5-50 ms, but I've seen up to -2.0 sec).

**Around 9 out of 10 of those events generate the core warning about the unknown listener**.
Filtering with `and self._last_scheduled_purge > dt_util.utcnow()` removes them (maybe coulb be necessary a little extra margin `dt_util.utcnow() + C`)

Reply:
You're right about `async_track_point_in_utc_time`. But the debouncer is guarding for it being called more than once. As long as the refresh hasn't executed, it won't schedule a new timer with the changes from #32161: https://github.com/azogue/home-assistant/blob/ec7c13300f02c266ffde1d5ad2e8ca3793b0cc14/homeassistant/helpers/debounce.py#L40-L44

Reply:
> But the debouncer is guarding for it being called more than once. As long as the refresh hasn't executed, it won't schedule a new timer

Yes, but the issue happens anyway (I tried that combination).

* With the changes in #32161 and debouncer set with 1s, it is true that the scheduling rate is maxed to 1 Hz, as the updater only runs once each second, tops.
* BUT, **the collision is between an update going on** (checking `if self._update_listener is None` in the way), **and an _unbounced callback_ scheduled a long time before**. Inside the callback, the call to update is debounced, but `async_track_point_in_utc_time` is not calling that directly.
* I suppose we could wrap it so the debouncer function IS the current callback, but anyway, we would be in the same position, as the point is that some listener can get caught in the middle between its unsub and the task running.

In the end, with simple logic to forget about listeners that may be running at the same moment, the issue is fixed, without any debouncer help

Reply:
The debouncer has a lock to prevent from breaking when being called when a run is in progress. 

And yes, there should be no unbounced callback. It should all go via the debouncer instance for it to work.

Reply:
> The debouncer has a lock to prevent from breaking when being called when a run is in progress.

That I now, and it works as expected.

> And yes, there should be no unbounced callback. It should all go via the debouncer instance for it to work.

It was also my thought. But reality always kicks in :(

I think I do not entirely follow you, as IMHO if there is a simple solution without debouncer, that should be preferable over a more complex one.

But, anyway, I'm back with the debouncer to check the behavior with debouncing the callback, and **it does not work**. In fact, it works much worse than before:
- Even if the _callback_ is debounced, the problem resides in the time between `async_unsub()` and the now debounced _action_ inside `hass.async_run_job(action, now)` (in the `async_track_point_in_utc_time` logic).
- Also, this does not avoid the fact that collisions are going to happen between incoming samples and scheduled operations, so canceling listeners needs to be done with care (:= by looking to their timestamps).
- With these changes (cooldown 1s, immediate=True, no timestamp control, callback IS debounced), everything breaks quickly, even in the dev machine. I got the core warning in debug with that:

```log
10:40:01.002 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:01.001008+00:00 and last_sched=2020-03-06 09:40:00.006207+00:00; #35
10:40:01.003 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#36)
10:40:01.003 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:01.003204+00:00(0:02:00)
10:40:02.007 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:01.001008+00:00 and last_sched=2020-03-06 09:40:03.002585+00:00; #36
10:40:02.008 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#37)
10:40:02.008 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:02.008718+00:00(0:02:00)
10:40:04.004 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:04.003315+00:00 and last_sched=2020-03-06 09:40:03.002585+00:00; #37
10:40:04.005 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#38)
10:40:04.005 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:04.005818+00:00(0:02:00)
10:40:05.009 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:04.003315+00:00 and last_sched=2020-03-06 09:40:05.003416+00:00; #38
10:40:05.010 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#39)
10:40:05.010 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:05.010232+00:00(0:02:00)
10:40:07.005 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#40)
10:40:07.006 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:07.006028+00:00(0:02:00)
10:40:08.009 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#41)
10:40:08.009 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:08.009793+00:00(0:02:00)
```
And just inside that last update nº41, the core warning appeared, trying to cancel the current listener, set to 2020-03-06 09:40:08.003434+00:00 !!, but this time I got in in debug :)

What happened there? Let's see:
- At 09:40:08 UTC, 2 things happened:
  * A new sample is received
  * A scheduled update is called
- Murphy's Law makes the incoming sample be first then the scheduled call, so the debouncer runs that one.
- Meanwhile, `async_track_point_in_utc_time` has activated the listener, done the unsub, and put the task in the queue, but that one is not going to run because of the debouncer (and that is the one that removes the handler with `self._update_listener = None`.
- When the update operation running at 09:40:08 UTC (nº41) tries to make a new schedule, sees that there is one already, and cancels it, **finally raising the core warning** about the unknown listener.
- In fact, with this configuration, the warning appears A LOT, as most of the scheduled operations (that turn off the listener flag) are aborted with the debouncer.
- **There are more scary things in the logs above**. Looks like some listeners are being called **more than once** (yes, I know the 'run' flag should prevent that, that is why I said _scary_):
  * `async_track_point_in_utc_time` pass the `now` arg to action (I had to add `*args` to the debouncer to make this combination work)
  * I'm using this to know inside the updater who's the caller (incoming sample does not add 'now' argument)
  * As I was going a little crazy :), I introduced an internal counter in the sensor, to track the calls (the _#i_ in logs)
  * Also, I included ms in log timestamps for richer detail (they are in local time, so 10:40 is 9:40 UTC)

  At 09:40:01 UTC, it seems like that that _tick_ activates 2 listeners, as update is called twice with the same 'now' timestamp (but they run separated with the 1s cooldown):
  
  ```log
  10:40:01.002 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:01.001008+00:00 and last_sched=2020-03-06 09:40:00.006207+00:00; #35
  10:40:01.003 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#36)
  10:40:02.007 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:01.001008+00:00 and last_sched=2020-03-06 09:40:03.002585+00:00; #36
  10:40:02.008 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#37)
  ```
  
  The second one has a different 'last_sched' date, 09:40:03 UTC, as it refers to the new schedule created in the first call nº36, which is in the future.

So in conclusion, as I see it:
- there is no good way to solve this issue with a debouncer/throttler with a window of ~1s, using `async_track_point_in_utc_time` for the purge scheduling.
- A debouncer with a much shorter cooldown is a very good tool for burst control, which is also happening here sometimes, but that is not the root cause for the issue to solve here.
- With simple timestamp logic to avoid excessive listener re-scheduling and collision, the issue totally disappears, and as a collateral benefit, the stress in the event bus is reduced (I noted around 1-2% CPU usage reduction in the RPI, from 11-12 to 9-10 %, but that is hard to prove or even be sure about)

Reply:
Code modifications for the analysis above are:

```python
# Debouncer config
_DEBOUNCING_COOLDOWN = 1.0
debounced_updater = Debouncer(
    hass, _LOGGER, cooldown=_DEBOUNCING_COOLDOWN, immediate=True
)
self._debounced_updater.function = self._async_debounced_update

# debounced function
async def _async_debounced_update(self, *args):
    """
    Debounced call to update entity.

    To process stats just once (in the end) when a burst of samples is received.
    """
    if args:
        # The call is coming from a scheduled update operation
        _LOGGER.warning(
            "%s: executing scheduled update with now=%s and last_sched=%s; #%d",
            self.entity_id,
            args[0],
            self._last_scheduled_purge,
            self._debug_update_counter,
        )
        self._update_listener = None
    await self.async_update_ha_state(True)

# Incoming samples:
self._add_state_to_queue(new_state)
await self._debounced_updater.async_call()

# Update method
async def async_update(self):
    """Get the latest data and updates the states."""
    self._debug_update_counter += 1
    _LOGGER.debug("%s: updating statistics (#%d)", self.entity_id, self._debug_update_counter)

    if self._max_age is not None:
        self._purge_old()

    self.count = len(self.states)

    if not self.is_binary:
        self._process_statistics()

    next_to_purge_timestamp = self._next_to_purge_timestamp()
    if next_to_purge_timestamp:
        if self._update_listener is not None:
            self._update_listener()
            self._update_listener = None

        self._last_scheduled_purge = next_to_purge_timestamp

        self._update_listener = async_track_point_in_utc_time(
            self.hass, self._debounced_updater.async_call, next_to_purge_timestamp
        )

# In debounce.py I had to add *args everywhere to make it work with async_track_point_in_utc_time
async def async_call(self, *args: Any) -> None:
   # ...
```

I will not push this, as that is an unknown listener warning factory :)

Reply:
_Outside the context of this PR/issue, as it is not related with the issue, some thoughts, @balloob:_

One of the things this PR evolution has convinced me is that **`async_track_point_in_utc_time` needs some change** because it is being used wrong here, IMO, and probably in more places.
- As the _clock tick_ is of 1Hz, it has no sense to program a listener in _t=1.2 s_, as the listener is going to run in _t=1.00x s_ or _t=2.00x s_ (probably the latter).
- In that context, all calls to `async_track_point_in_utc_time` should be done with exact seconds, always minimizing the delay between the expected time and the real one (activation is done when `now >= point_in_time`).
- Quick solution would be to **round** or **ceil** the given `point_in_time` inside `async_track_point_in_utc_time`, to discard the sub-second precission, but deciding between round and ceil is not clear.